.github/workflows/check-api-changes.yaml

```diff
@@ -0,0 +1,101 @@
+
+      - name: Checkout main of lifecycle-manager
+        uses: actions/checkout@v4
+        with:
+          ref: main
+
+      - name: Copy main CRDs to temp dir
+        run: mkdir -p /tmp/main-branch-crds && cp ./config/crd/bases/* /tmp/main-branch-crds/
+
+      - name: Compare names and number of CRD files
+        working-directory: /tmp
+        run: |
+          main_files=$(ls ./main-branch-crds)
+          pr_files=$(ls ./pr-crds)
+          
+          if [[ "$pr_files" != "$main_files" ]]; then
+            echo "Number of CRD files do not match"
+            echo "Differences between main-branch-crds and pr-crds:"
+            diff -r main-branch-crds pr-crds || true
+            exit 1
+          fi
+          echo "Number of CRD files match"
+
+      - name: Compare names of API versions in CRD files
+        working-directory: /tmp
+        run: |
+          compare_versions() {
+            main_file=$1
+            pr_file=$2
+
+            main_versions=$(yq e '.spec.versions[].name' "$main_file" | sort | tr '\n' ' ')
+            pr_versions=$(yq e '.spec.versions[].name' "$pr_file" | sort | tr '\n' ' ')
+
+            if [[ "$pr_versions" != "$main_versions" ]]; then
+              echo "API versions do not match in $pr_file"
+              echo "Versions on main: $main_versions"
+              echo "Versions in PR: $pr_versions"
+              exit 1
+            fi
+          }
+
+          for file in $(ls ./pr-crds); do
+            compare_versions "./main-branch-crds/$file" "./pr-crds/$file"
+          done
+
+          echo "Name of API versions match in all CRD files"
+
+      - name: Compare API schemas in CRD files
+        working-directory: /tmp
+        run: |
+          compare_schemas() {
+            main_file=$1
+            pr_file=$2
+            version=$3
+
+            main_schema_file="/tmp/main-crds-schema-${version}.yaml"
+            pr_schema_file="/tmp/pr-crds-schema-${version}.yaml"
+
+            yq '.spec.versions[] | select(.name == "'$version'") | .schema.openAPIV3Schema.properties | del(.. | select(has("description")).description)' "$main_file" > "$main_schema_file"
+            yq '.spec.versions[] | select(.name == "'$version'") | .schema.openAPIV3Schema.properties | del(.. | select(has("description")).description)' "$pr_file" > "$pr_schema_file"
+
+            if ! diff "$main_schema_file" "$pr_schema_file" ; then
+              echo "Schema differences found for version $version in $main_file and $pr_file"
+              exit 1
+            fi
+          }
+
+          for file in $(ls ./pr-crds); do
+            main_file="./main-branch-crds/$file"
+            pr_file="./pr-crds/$file"
+            versions=$(yq e '.spec.versions[].name' "$pr_file")
+
+            for version in $versions; do
+              compare_schemas "$main_file" "$pr_file" "$version"
+            done
+          done
+
+          echo "API schemas match in all CRD files"
```

Comment:
There are a couple of things that need to be changed here:

- a) we need to compare different versions of the APIs, e.g. `v1beta1` against `v1beta2` (see also the example here https://github.com/kyma-project/lifecycle-manager/issues/1247#issuecomment-2066782823). The reason is that we need to ensure that when K8s stores one version of the CRD in etcd, it must be able to translate this version to other served versions. Hence we want to be alerted when the API changes between versions so that we can check if we need to backport something to the older API. And for the known changes and okay ones, we need to do the pre-processing (e.g., as in the example remove the `sync` field from `v1beta1` as this is a known and okay change).
Since we may have more than two API versions in the future, I would propose that we take the latest version and compare it against the predecessor, then take the predecessor and compare it against its predecessor and so on. With this change, we then should also be able get rid of the `Compare names of API versions in CRD files` step (as we may introduce new versions, we just need to take care about the compatiblity), and we also may get rid of the label to accept a change (accepting a change is done by pre-processing a known difference so that it doesn't yield a difference output)
- b) use `dyff` instead of `diff`. Apologies, this was discussed in the team when looking at https://github.com/kyma-project/lifecycle-manager/issues/1247#issuecomment-2066782823, but not stated clearly as part of this ticket. Main reasons are the license, better formatting of the output, and the option to ignore descriptions.
- c) I would propose to move the preprocessing to another step if possible. This step should be named something like "preprocessing known differences" so it is obvious that we have a set of differences that we need to pre-process and where to add future pre-processing

Reply:
oh I guess I misunderstood the issue a bit. I wrote the workflow to compare crd changes between main and pr. I will modify it for intra-crd version comparison.